editor/src/clj/editor/code/view.clj

```diff
@@ -1425,37 +1527,56 @@
-                                            (do (accept-suggestion! view-node)
-                                                [false false])
-
-                                            (and (= "(" typed) (= :function (:type selected-suggestion)))
-                                            (do (accept-suggestion! view-node)
-                                                [false false])
-
-                                            (and (= "." typed) (= :namespace (:type selected-suggestion)))
-                                            (do (accept-suggestion! view-node)
-                                                [true true])
-
-                                            (data/typing-deindents-line? grammar lines cursor-ranges typed)
-                                            [true false]
+        [insert-typed show-suggestions]
+        (cond
+          (and selected-suggestion
+               (or (= "\r" typed)
+                   (let [commit-characters (or (:commit-characters selected-suggestion)
+                                               (get-in grammar [:commit-characters (:type selected-suggestion)]))]
+                     (contains? commit-characters typed))))
+          (let [insertion (code-completion/insertion selected-suggestion)]
+            (do (accept-suggestion! view-node insertion)
+                [#_insert-typed (not
+                                  ;; exclude typed when...
+                                  (or (= typed "\r")
+                                      ;; At this point, we know we typed a commit character.
+                                      ;; If there are tab stops, and we typed a character
+                                      ;; before the tab stop, we assume the commit character
+                                      ;; is a shortcut for accepting a completion and jumping
+                                      ;; into the tab stop, e.g. foo($1) + "(" => don't
+                                      ;; insert. Otherwise, we insert typed, e.g.:
+                                      ;; - foo($1) + "{" => the typed character is expected
+                                      ;;   to be inside the tab stop, for example, when foo
+                                      ;;   expects a hash map
+                                      ;; - vmath + "." => the typed character is expected
+                                      ;;   to be after the snippet.
+                                      (when-let [^long i (->> insertion :tab-triggers first :ranges first first)]
+                                        (when (pos? i)
+                                          (= typed (subs (:insert-string insertion) (dec i) i))))))
+                 #_show-suggestions (not
+                                      ;; hide suggestions when entering new scope
+                                      (#{"[" "(" "{"} typed))]))
```

Comment:
No, those are not a part of the experiment but are used for clarity. I'll replace those with ;;-style comments :) 
I worry that introducing even more let bindings with same names will only add confusion, since this vector is already used for let bindings...